telemetry/attributes.go

```diff
@@ -29,45 +31,46 @@ func vetAttributes(attributes map[string]interface{}, errorLogger func(map[strin
 		}
 	}
 	if valid {
-		return attributes
+		return attributes, nil
 	}
 	// Note that the map is only copied if elements are to be removed to
 	// improve performance.
 	validAttributes := make(map[string]interface{}, len(attributes))
+	var errStrs []string
 	for key, val := range attributes {
 		if attributeValueValid(val) {
 			validAttributes[key] = val
-		} else if nil != errorLogger {
-			errorLogger(map[string]interface{}{
-				"err": fmt.Sprintf(`attribute "%s" has invalid type %T`, key, val),
-			})
+		} else {
+			errStrs = append(errStrs, fmt.Sprintf(`attribute "%s" has invalid type %T`, key, val))
 		}
 	}
-	return validAttributes
+	return validAttributes, errors.New(strings.Join(errStrs, ","))
 }
 
-type commonAttributes struct {
+// CommonAttributes defines a shared set of attributes.
+type CommonAttributes struct {
 	RawJSON json.RawMessage
 }
 
-func (ca *commonAttributes) Type() string {
+// Type returns the type of data contained in this PayloadEntry.
+func (ca *CommonAttributes) Type() string {
 	return "attributes"
 }
 
-func (ca *commonAttributes) Bytes() []byte {
+// Bytes returns the json serialized bytes of the PayloadEntry.
+func (ca *CommonAttributes) Bytes() []byte {
 	return ca.RawJSON
 }
 
-func newCommonAttributes(attributes map[string]interface{}, errorLogger func(map[string]interface{})) (*commonAttributes) {
-	attrs := vetAttributes(attributes, errorLogger)
+// NewCommonAttributes vets the attributes map and converts it to a CommonAttributes.
+func NewCommonAttributes(attributes map[string]interface{}) (*CommonAttributes, error) {
```

Comment:
I changed the contract of this to be more idiomatic for callers outside of the telemetry sdk. Internal callers can still check for the error return and pass it to an error logger, but external callers don't have to.

Reply:
Perhaps this method and the corresponding CommonAttributes shouldn't be exported, and we instead rely on a builder that produces a `PayloadEntry`.

What could do something more like:
```
spanCommonBlock := NewSpanCommonBlockBuilder()
    .WithCommonAttributes(attributeMap, errorLogger)
    .Build()
```
There could also be a variant without the errorLogger too, if that makes sense.

This would allow us to handle metrics by doing something like the following where the different builder methods are optional:
```
metricCommonBlock := NewMetricCommonBlock()
    .WithCommonAttributes(attributesMap, errorLogger)
    .WithTimestamp(timestamp)
    .WithInterval(interval)
    .Build()
```
Then you wouldn't need to change SpanBatch to include the common block. The common block is just an optional PayloadEntry that can be passed when building a request.

Reply:
So I initially didn't export `CommonAttributes`, but the linter yelled at me and said that having exported methods that return types that aren't exported are confusing to work with. Not sure I agree, but that's why I ended up exporting it.

Having SpanBatch include a common block means that the code mirrors the structure of the json:
- A request body is composed of 1 or more span batches
- A span batch is composed of an optional common block and 1 or more spans

If we broke this symmetry and instead said that you pass the common block when building a request, then we'd have to send multiple requests in the collector exporter whenever processing a request with multiple resources or instrumentation libraries. The hierarchy of the data coming into the collector is: `Resources -> InstrumentationLibraries -> Spans`. So we need to be able to generate a span batch with a unique set of common attributes for each incoming Resource - InstrumentationLibrary tuple.